### PR TITLE
Refactoring Task Infrastructure

### LINKED ISSUE #40954: Migrate To Google Wireit
There is currently a spike in https://github.com/woocommerce/woocommerce/pull/39504 for a migration from Turborepo to Wireit. With https://github.com/woocommerce/woocommerce/issues/40953 and https://github.com/woocommerce/woocommerce/issues/40952 in mind, we will also include the migration to Wireit with those changes.

### Acceptance Criteria

* Replace Turborepo with Wireit.
* Update documentation to reflect this change.

### PR DESCRIPTION
**Pull request by ObliviousHarmony**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This pull request makes a number of changes to our task orchestration infrastructure in order to introduce consistency and significantly improve cache performance. Generally speaking, from the user-perspective, there should be very little changed by this pull request.

**Note: I chose to batch the changes in this PR so that it will have a minimal impact on developer experience by being merged at the same time.**

Closes https://github.com/woocommerce/woocommerce/issues/40954, https://github.com/woocommerce/woocommerce/issues/40952, https://github.com/woocommerce/woocommerce/issues/40953.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Verify that CI is passing. All of the changes in this PR apply to CI as well.
2. `pnpm --filter='@woocommerce/plugin-woocommerce' build`
3. `pnpm --filter='@woocommerce/plugin-woocommerce' lint`
4. `pnpm --filter='@woocommerce/plugin-woocommerce' build`
5. Explore! Play around with different commands and see how things work out.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by rrennick** [comment]
@ObliviousHarmony There's lots of merge conflicts with this. Is this something that you will need to resolve just prior to merging?

**Comment by ObliviousHarmony** [comment]
I fixed the conflicts @rrennick. The joy of such a large PR :smile:

**Comment by rrennick** [comment]
I noticed that linting logged errors in the `build` folder for blocks:

```
FILE: build/woocommerce/packages/woocommerce-blocks/build/stock-filter.asset.php
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 1 LINE
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 1 | ERROR | [ ] Missing file doc comment (Squiz.Commenting.FileComment.Missing)
 1 | ERROR | [x] Missing space after array opener. (WordPress.Arrays.ArrayDeclarationSpacing.NoSpaceAfterArrayOpener)
 1 | ERROR | [x] Missing space before array closer. (WordPress.Arrays.ArrayDeclarationSpacing.NoSpaceBeforeArrayCloser)
 1 | ERROR | [x] When a multi-item array uses associative keys, each value should start on a new line.
   |       |     (WordPress.Arrays.ArrayDeclarationSpacing.AssociativeArrayFound)
```

These probably don't exist in the blocks repo but post merge these files will exist in a local dev environment and should be excluded from linting. 

As it is, should the lint be processing files in `plugins/woocommerce/packages`at all as the contents of the folder is pulled in from other repositories?

**Comment by rrennick** [comment]
~~The pnpm commands in https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/tests/performance/README.md have not been updated.~~

Edit: I'd accidentally switched back to trunk.

File: packages/js/e2e-environment/external.md [hunk 1]
```diff
@@ -38,7 +38,7 @@ Again, if you don't have shell access to your test site, through WP Admin ensure
 3. You have an admin user set up (if their credentials differ from u/ `admin` and p/ `password` be sure to update `/plugins/woocommerce/tests/e2e/config/default.json`)
 4. You have a customer user set up named 'Jane Smith'. This user should be a `subscriber` and again make sure their username and password are reflected in `/plugins/woocommerce/tests/e2e/config/default.json`.
 
-You should then be able to run the e2e tests by running `pnpm run e2e --filter=woocommerce`.
+You should then be able to run the e2e tests by running `pnpm run e2e --filter=@woocommerce/plugin-woocommerce`.
```
Comment (rrennick) [inline packages/js/e2e-environment/external.md]: Adding a note that I created #41649 to address the use of monorepo specific commands in these `.md` files.

File: plugins/woocommerce/tests/performance/README.md [hunk 2]
```diff
@@ -57,8 +57,8 @@ Before using the tests a test environment is needed to run against.
 We first spin up an environment using `wp-env` and configure that environment with the necessary plugins and data using the Initialization Script [`init-sample-products.sh`](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/tests/performance/bin/init-sample-products.sh) that will set up a shop with sample products imported and the shop settings (payment method, permalinks, address etc) needed for the tests already set. It is recommended using this to just see the tests in action.
 
 ```sh
-pnpm run env:dev --filter=woocommerce
-pnpm run env:performance-init --filter=woocommerce
+pnpm run env:dev --filter=@woocommerce/plugin-woocommerce
```
Comment (rrennick) [inline plugins/woocommerce/tests/performance/README.md]: This command is loading e2e-environment instead of wp-env.

Comment (ObliviousHarmony) [reply plugins/woocommerce/tests/performance/README.md]: That script runs `"env:dev": "pnpm wp-env start --update"` and should be setting up `wp-env`.

Comment (rrennick) [reply plugins/woocommerce/tests/performance/README.md]: Also, curious about why 3 `plugins/woocommerce/tests/*/README.md` were deleted.

Comment (rrennick) [reply plugins/woocommerce/tests/performance/README.md]: > That script runs "env:dev": "pnpm wp-env start --update" and should be setting up wp-env.

`run` needs to be removed.

```suggestion
pnpm env:dev --filter=@woocommerce/plugin-woocommerce
```

File: plugins/woocommerce/tests/performance/README.md [hunk 3]
```diff
@@ -57,8 +57,8 @@ Before using the tests a test environment is needed to run against.
...
+pnpm run env:performance-init --filter=@woocommerce/plugin-woocommerce
```
Comment (rrennick) [inline plugins/woocommerce/tests/performance/README.md]: `run` should be removed here as well.

**Review by rrennick — COMMENTED** [review]
@jonathansadowski I've left a few comments and created some issues that weren't directly related to this PR. I did not find any functional issues in my testing of build and testing scripts. I did not test any of the release related scripting.

@ObliviousHarmony Nice work. Other than the comments I left it would be great to have this rebased to current `trunk` and have CI pass before merging. IIRC, this branch is about 150 commits behind trunk.

**Comment by jonathansadowski** [comment]
> @jonathansadowski I've left a few comments and created some issues that weren't directly related to this PR. I did not find any functional issues in my testing of build and testing scripts. I did not test any of the release related scripting.

I've tested some of the release-related things... namely building release zips, as well as the tooling used for code-freeze, none of that broke, though I wouldn't expect the latter to.

I also didn't discover any functional issues, though I still have a list of things I wanted to check, and I'm not sure I will ahead of the holiday weekend, but that list is entirely related to the changes to `.github/actions/setup-woocommerce-monorepo/action.yml`, mostly because I recall previous changes to this action causing some unforeseen issues.

I didn't notice any functional issues with anything I've checked so far, and the things I wanted to check in the changes to `setup-woocommerce-monorepo` all appeared to be non-problematic at first glance (things like switching from string literals to boolean with the input parameters), so I think I'd be comfortable with this PR if like Ron said this were rebased to current trunk and CI were passing.

**Pull request merged**